src/sentry/migrations/0348_issue_alert_fallback.py

```diff
@@ -0,0 +1,84 @@
+# Generated by Django 2.2.28 on 2023-01-04 19:14
+
+import logging
+
+from django.db import migrations, transaction
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+
+def set_issue_alert_fallback(rule, fallthrough_choice):
+    actions = rule.data.get("actions", [])
+    new_actions = []
+    for action in actions:
+        id = action.get("id")
+        target_type = action.get("targetType")
+        if id == "sentry.mail.actions.NotifyEmailAction" and target_type == "IssueOwners":
+            action.update({"fallthroughType": fallthrough_choice})
+            new_actions.append(action)
+
+    rule.save()
+
+
+def migrate_project_ownership_to_issue_alert_fallback(project, ProjectOwnership, Rule):
+    with transaction.atomic():
+        # Determine whether this project has a fallback setting.
+        fallthrough_choice = None
+        try:
+            ownership = ProjectOwnership.objects.get(project=project)
+            fallthrough_choice = "AllMembers" if ownership and ownership.fallthrough else "NoOne"
+        except ProjectOwnership.DoesNotExist:
+            # Projects without fallbacks will be assigned the new "ActiveMembers" default.
+            fallthrough_choice = "ActiveMembers"
+
+        # We only migrate rules that are not pending deletion.
+        for rule in Rule.objects.filter(project=project, status=0):
+            set_issue_alert_fallback(rule, fallthrough_choice)
+
+        project.save()
+
+
+def migrate_to_issue_alert_fallback(apps, schema_editor):
+    Project = apps.get_model("sentry", "Project")
+    ProjectOwnership = apps.get_model("sentry", "ProjectOwnership")
+    Organization = apps.get_model("sentry", "Organization")
+    Rule = apps.get_model("sentry", "Rule")
+
+    # We migrate a project at a time, but we prefer to group by org so that for the
+    # most part an org will see the changes all at once.
+    for org in RangeQuerySetWrapperWithProgressBar(Organization.objects.filter(status=0)):
+        # We only migrate projects that are not pending deletion.
+        for project in Project.objects.filter(organization=org, status__in=[0, 1]):
+            try:
+                migrate_project_ownership_to_issue_alert_fallback(project, ProjectOwnership, Rule)
+            except Exception:
+                # If a project fails we'll just log and continue. We shouldn't see any
+                # failures, but if we do we can analyze them and run a new migration.
+                logging.exception(f"Error migrating project {project.id}")
```

Comment:
We probably want to retry the migration for failed projects if this happens. [This migration ](https://github.com/getsentry/sentry/pull/18330/files)uses a project-level flag for this. 

@scttcper do we want to do something similar? We don't have too many [project flags](https://github.com/getsentry/sentry/pull/18330/files) so not sure if this is an appropriate use case

Reply:
How would you do this?
Would you update a flag for a project so you can check for it later?

Reply:
Ah, that makes sense! I think we can work with ops to handle any failures - thanks!